File: js/generator.js
const defaultUnambiguousCharacters = "abcdefghijkmnopqrstuvwxyzABCDEFGHIJKLMNPQRSTUVWXYZ0123456789";

const defaultNumberOfCharactersForClassicPassword = 12;
const defaultClassicPasswordLength = 15;

const defaultNumberOfCharactersForMoreTypeablePassword = 18;
const defaultMoreTypeablePasswordLength = 20;
const defaultAllowedNumbers = "0123456789";
const defaultAllowedLowercaseConsonants = "bcdfghjkmnpqrstvwxz";
const defaultAllowedLowercaseVowels = "aeiouy";

var PasswordGenerationStyle = {
    Classic: 1,
    ClassicWithoutDashes: 2,
    MoreTypeable: 3,
    MoreTypeableWithoutDashes: 4,
};

function randomNumberWithUniformDistribution(range)
{
    // Based on the algorithn described in https://pthree.org/2018/06/13/why-the-multiply-and-floor-rng-method-is-biased/
    var max = Math.floor(2**32 / range) * range;
    do {
        var x = window.crypto.getRandomValues(new Uint32Array(1))[0];
    } while (x >= max);

    return (x % range);
}

function randomConsonant()
{
    var index = randomNumberWithUniformDistribution(defaultAllowedLowercaseConsonants.length);
    return defaultAllowedLowercaseConsonants[index];
}

function randomVowel()
{
    var index = randomNumberWithUniformDistribution(defaultAllowedLowercaseVowels.length);
    return defaultAllowedLowercaseVowels[index];
}

function randomNumber()
{
    var index = randomNumberWithUniformDistribution(defaultAllowedNumbers.length);
    return defaultAllowedNumbers[index];
}

function randomSyllable()
{
    return randomConsonant() + randomVowel() + randomConsonant();
}

function _moreTypeablePassword()
{
    var password = randomSyllable() + randomSyllable() + randomSyllable() + randomSyllable() + randomSyllable() + randomConsonant() + randomVowel();
    var length = password.length;
    while (true) {
        var index = randomNumberWithUniformDistribution(length);
        var lowercaseChar = password.charAt(index);
        if (lowercaseChar === "o")
            continue;

        var uppercaseChar = lowercaseChar.toUpperCase();
        password = password.substr(0, index) + uppercaseChar + password.substr(index + 1);

        var numberPos = randomNumberWithUniformDistribution(5);
        var passwordSegment1 = password.substr(0, 6);
        var passwordSegment2 = password.substr(6, 6);
        var passwordSegment3 = password.substr(12, 5);
        switch (numberPos) {
        case 0:
            return passwordSegment3 + randomNumber() + passwordSegment1 + passwordSegment2;
        case 1:
            return passwordSegment1 + randomNumber() + passwordSegment3 + passwordSegment2;
        case 2:
            return passwordSegment1 + passwordSegment3 + randomNumber() + passwordSegment2;
        case 3:
            return passwordSegment1 + passwordSegment2 + randomNumber() + passwordSegment3;
        case 4:
            return passwordSegment1 + passwordSegment2 + passwordSegment3 + randomNumber();
        }
    }
}

function _classicPassword(numberOfRequiredRandomCharacters, allowedCharacters)
{
    var length = allowedCharacters.length;
    var randomCharArray = Array(numberOfRequiredRandomCharacters);
    for (var i = 0; i < numberOfRequiredRandomCharacters; i++) {
        var index = randomNumberWithUniformDistribution(length);
        randomCharArray[i] = allowedCharacters[index];
    }
    return randomCharArray.join("");
}

function _passwordHasNotExceededConsecutiveCharLimit(password, consecutiveCharLimit)
{
    var longestConsecutiveCharLength = 1;
    var firstConsecutiveCharIndex = 0;
    // Both "123" or "abc" and "321" or "cba" are considered consecutive.
    var isSequenceAscending;
    for (var i = 1; i < password.length; i++) {
        var currCharCode = password.charCodeAt(i);
        var prevCharCode = password.charCodeAt(i-1);
        if (isSequenceAscending) {
            // If `isSequenceAscending` is defined, then we know that we are in the middle of an existing
            // pattern. Check if the pattern continues based on whether the previous pattern was
            // ascending or descending.
            if ((isSequenceAscending.valueOf() && currCharCode == prevCharCode + 1) || (!isSequenceAscending.valueOf() && currCharCode == prevCharCode - 1))
                continue;

            // Take into account the case when the sequence transitions from descending
            // to ascending.
            if (currCharCode == prevCharCode + 1) {
                firstConsecutiveCharIndex = i - 1;
                isSequenceAscending = Boolean(true);
                continue;
            }

            // Take into account the case when the sequence transitions from ascending
            // to descending.
            if (currCharCode == prevCharCode - 1) {
                firstConsecutiveCharIndex = i - 1;
                isSequenceAscending = Boolean(false);
                continue;
            }

            isSequenceAscending = null;
        } else if (currCharCode == prevCharCode + 1) {
            isSequenceAscending = Boolean(true);
            continue;
        } else if (currCharCode == prevCharCode - 1) {
            isSequenceAscending = Boolean(false);
            continue;
        }

        var currConsecutiveCharLength = i - firstConsecutiveCharIndex;
        if (currConsecutiveCharLength > longestConsecutiveCharLength)
            longestConsecutiveCharLength = currConsecutiveCharLength;

        firstConsecutiveCharIndex = i;
    }

    if (isSequenceAscending) {
        var currConsecutiveCharLength = password.length - firstConsecutiveCharIndex;
        if (currConsecutiveCharLength > longestConsecutiveCharLength)
            longestConsecutiveCharLength = currConsecutiveCharLength;
    }

    return longestConsecutiveCharLength <= consecutiveCharLimit;
}

function _passwordHasNotExceededRepeatedCharLimit(password, repeatedCharLimit)
{
    var longestRepeatedCharLength = 1;
    var lastRepeatedChar = password.charAt(0);
    var lastRepeatedCharIndex = 0;
    for (var i = 1; i < password.length; i++) {
        var currChar = password.charAt(i);
        if (currChar === lastRepeatedChar)
            continue;

        var currRepeatedCharLength = i - lastRepeatedCharIndex;
        if (currRepeatedCharLength > longestRepeatedCharLength)
            longestRepeatedCharLength = currRepeatedCharLength;

        lastRepeatedChar = currChar;
        lastRepeatedCharIndex = i;
    }
    return longestRepeatedCharLength <= repeatedCharLimit;
}

function _passwordContainsRequiredCharacters(password, requiredCharacterSets)
{
    var requiredCharacterSetsLength = requiredCharacterSets.length;
    var passwordLength = password.length;
    for (var i = 0; i < requiredCharacterSetsLength; i++) {
        var requiredCharacterSet = requiredCharacterSets[i];
        var hasRequiredChar = false;
        for (var j = 0; j < passwordLength; j++) {
            var char = password.charAt(j);
            if (requiredCharacterSet.indexOf(char) !== -1) {
                hasRequiredChar = true;
                break;
            }
        }
        if (!hasRequiredChar)
            return false;
    }
    return true;
}

function _defaultRequiredCharacterSets()
{
    return [ "abcdefghijklmnopqrstuvwxyz", "ABCDEFGHIJKLMNOPQRSTUVWXYZ", "0123456789" ];
}

function _stringContainsAllCharactersInString(string1, string2)
{
    var length = string2.length;
    for (var i = 0; i < length; i++) {
        var character = string2.charAt(i);
        if (string1.indexOf(character) === -1)
            return false;
    }
    return true;
}

function _stringsHaveAtLeastOneCommonCharacter(string1, string2)
{
    var string2Length = string2.length;
    for (var i = 0; i < string2Length; i++) {
        var char = string2.charAt(i);
        if (string1.indexOf(char) !== -1)
            return true;
    }

    return false;
}

function _canUseMoreTypeablePasswordFromRequirements(minPasswordLength, maxPasswordLength, allowedCharacters, requiredCharacterSets)
{
    if ((minPasswordLength > defaultMoreTypeablePasswordLength) || (maxPasswordLength && maxPasswordLength < defaultMoreTypeablePasswordLength))
        return false;

    if (allowedCharacters && !_stringContainsAllCharactersInString(allowedCharacters, defaultUnambiguousCharacters))
        return false;

    var requiredCharacterSetsLength = requiredCharacterSets.length;
    if (requiredCharacterSetsLength > defaultMoreTypeablePasswordLength)
        return false;

    // FIXME: This doesn't handle returning false to a password that requires two or more special characters.
    let numberOfDigitsThatAreRequired = 0;
    let numberOfUpperThatAreRequired = 0;
    for (let requiredCharacterSet of requiredCharacterSets) {
        if (requiredCharacterSet === "0123456789")
            numberOfDigitsThatAreRequired++;
        if (requiredCharacterSet === "ABCDEFGHIJKLMNOPQRSTUVWXYZ")
            numberOfUpperThatAreRequired++;
    }

    if (numberOfDigitsThatAreRequired > 1)
        return false;
    if (numberOfUpperThatAreRequired > 1)
        return false;

    var defaultUnambiguousCharactersPlusDash = defaultUnambiguousCharacters + "-";
    for (var i = 0; i < requiredCharacterSetsLength; i++) {
        var requiredCharacterSet = requiredCharacterSets[i];
        if (!_stringsHaveAtLeastOneCommonCharacter(requiredCharacterSet, defaultUnambiguousCharactersPlusDash))
            return false;
    }

    return true;
}

function _passwordGenerationParametersDictionary(requirements)
{
    var minPasswordLength = requirements["PasswordMinLength"];
    var maxPasswordLength = requirements["PasswordMaxLength"];

    if (minPasswordLength > maxPasswordLength) {
        // Resetting invalid value of min length to zero means "ignore min length parameter in password generation".
        minPasswordLength = 0;
    }

    var allowedCharacters = requirements["PasswordAllowedCharacters"];

    var requiredCharacterArray = requirements["PasswordRequiredCharacters"];
    var requiredCharacterSets = _defaultRequiredCharacterSets();
    if (requiredCharacterArray) {
        var mutatedRequiredCharacterSets = [];
        var requiredCharacterArrayLength = requiredCharacterArray.length;
        for (var i = 0; i < requiredCharacterArrayLength; i++) {
            var requiredCharacters = requiredCharacterArray[i];
            if (_stringsHaveAtLeastOneCommonCharacter(requiredCharacters, allowedCharacters))
                mutatedRequiredCharacterSets.push(requiredCharacters);
        }
        requiredCharacterSets = mutatedRequiredCharacterSets;
    }

    var canUseMoreTypeablePassword = _canUseMoreTypeablePasswordFromRequirements(minPasswordLength, maxPasswordLength, allowedCharacters, requiredCharacterSets);
    if (canUseMoreTypeablePassword) {
        var style =  PasswordGenerationStyle.MoreTypeable;
        if (allowedCharacters && (allowedCharacters.indexOf("-") === -1))
            style = PasswordGenerationStyle.MoreTypeableWithoutDashes;

        return { "PasswordGenerationStyle": style };
    }

    // If requirements allow, we will generate the password in default format: "xxx-xxx-xxx-xxx".
    var style = PasswordGenerationStyle.Classic;
    var numberOfRequiredRandomCharacters = defaultNumberOfCharactersForClassicPassword;
    if (minPasswordLength && minPasswordLength > defaultClassicPasswordLength) {
        style = PasswordGenerationStyle.ClassicWithoutDashes;
        numberOfRequiredRandomCharacters = minPasswordLength;
    }

    if (maxPasswordLength && maxPasswordLength < defaultClassicPasswordLength) {
        style = PasswordGenerationStyle.ClassicWithoutDashes;
        numberOfRequiredRandomCharacters = maxPasswordLength;
    }

    if (allowedCharacters) {
        // We cannot use default format if dash is not an allowed character in the password.
        if (allowedCharacters.indexOf("-") === -1)
            style = PasswordGenerationStyle.ClassicWithoutDashes;
    } else
        allowedCharacters = defaultUnambiguousCharacters;

    // In default password format, we use dashes only as separators, not as symbols you can encounter at a random position.
    if (style == PasswordGenerationStyle.Classic)
        allowedCharacters = allowedCharacters.replace(/-/g, "");

    if (!requiredCharacterSets)
        requiredCharacterSets = _defaultRequiredCharacterSets();

    // If we have more requirements of the type "need a character from set" than the length of the password we want to generate, then
    // we will never be able to meet these requirements, and we'll end up in an infinite loop generating passwords. To avoid this,
    // reset required character sets if the requirements are impossible to meet.
    if (requiredCharacterSets.length > numberOfRequiredRandomCharacters) {
        requiredCharacterSets = null;
    }

    // Do not require any character sets that do not contain allowed characters.
    var requiredCharacterSetsLength = requiredCharacterSets.length;
    var mutatedRequiredCharacterSets = [];
    var allowedCharactersLength = allowedCharacters.length;
    for (var i = 0; i < requiredCharacterSetsLength; i++) {
        var requiredCharacterSet = requiredCharacterSets[i];
        var requiredCharacterSetContainsAllowedCharacters = false;
        for (var j = 0; j < allowedCharactersLength; j++) {
            var character = allowedCharacters.charAt(j);
            if (requiredCharacterSet.indexOf(character) !== -1) {
                requiredCharacterSetContainsAllowedCharacters = true;
                break;
            }
        }
        if (requiredCharacterSetContainsAllowedCharacters)
            mutatedRequiredCharacterSets.push(requiredCharacterSet);
    }
    requiredCharacterSets = mutatedRequiredCharacterSets;

    return {
        "PasswordGenerationStyle": style,
        "NumberOfRequiredRandomCharacters": numberOfRequiredRandomCharacters,
        "PasswordAllowedCharacters": allowedCharacters,
        "RequiredCharacterSets": requiredCharacterSets,
    };
}

function generatedPasswordMatchingRequirements(requirements)
{
    requirements = requirements ? requirements : {};

    var parameters = _passwordGenerationParametersDictionary(requirements);
    var style = parameters["PasswordGenerationStyle"];
    var numberOfRequiredRandomCharacters = parameters["NumberOfRequiredRandomCharacters"];
    var repeatedCharLimit = requirements["PasswordRepeatedCharacterLimit"];
    var allowedCharacters = parameters["PasswordAllowedCharacters"];
    var shouldCheckRepeatedCharRequirement = repeatedCharLimit ? true : false;

    while (true) {
        var password;
        switch (style) {
            case PasswordGenerationStyle.Classic:
            case PasswordGenerationStyle.ClassicWithoutDashes:
                password = _classicPassword(numberOfRequiredRandomCharacters, allowedCharacters);
                if (style === PasswordGenerationStyle.Classic)
                    password = password.substr(0, 3) + "-" + password.substr(3, 3) + "-" + password.substr(6, 3) + "-" + password.substr(9, 3);

                if (!_passwordContainsRequiredCharacters(password, parameters["RequiredCharacterSets"]))
                    continue;

                break;
            case PasswordGenerationStyle.MoreTypeable:
            case PasswordGenerationStyle.MoreTypeableWithoutDashes:
                password = _moreTypeablePassword();
                if (style === PasswordGenerationStyle.MoreTypeable)
                    password = password.substr(0, 6) + "-" + password.substr(6, 6) + "-" + password.substr(12, 6);

                if (shouldCheckRepeatedCharRequirement && repeatedCharLimit !== 1)
                    shouldCheckRepeatedCharRequirement = false;

                break;
        }

        if (shouldCheckRepeatedCharRequirement) {
            if (repeatedCharLimit >= 1 && !_passwordHasNotExceededRepeatedCharLimit(password, repeatedCharLimit))
                continue;
        }

        var consecutiveCharLimit = requirements["PasswordConsecutiveCharacterLimit"]
        if (consecutiveCharLimit) {
            if (consecutiveCharLimit >= 1 && !_passwordHasNotExceededConsecutiveCharLimit(password, consecutiveCharLimit))
                continue;
        }

        return password;
    }
}

File: pwgen-swift/pwgen/Generator.swift
//
//  Generator.swift
//  pwgen
//
//  Created by Max-Joseph on 01.07.23.
//  Originally ported from JavaScript code at https://developer.apple.com/password-rules/scripts/generator.js
//  but modified to be more robust and versatile.
//

import Foundation


struct Generator {
	enum Style {
		case random
		case nice
	}
	
	private typealias CharacterCollection = Collection<Character>
	
	
	static let defaultGroupSeparator: Character = "-"
	static let defaultAllowedCharacters: CharacterSet = .unambiguous
	static let defaultRequiredCharacterSets: [CharacterSet] = [.lower, .upper, .digit]
	
	private let style: Style
	private let numberOfCharacters: Int
	private let group: (size: Int, separator: Character)?
	private let characterPool: CharacterSet
	private let requiredCharacterSets: [CharacterSet]
	private let repeatedCharacterLimit: Int?
	private let consecutiveCharacterLimit: Int?
	
	init(
		style: Style,
		minLength: Int,
		grouped: Bool,
		groupSeparator: Character = Self.defaultGroupSeparator,
		allowedCharacters: CharacterSet = Self.defaultAllowedCharacters,
		requiredCharacterSets: [CharacterSet] = Self.defaultRequiredCharacterSets,
		repeatedCharacterLimit: Int? = nil,
		consecutiveCharacterLimit: Int? = nil
	) {
		self.style = style
		self.consecutiveCharacterLimit = consecutiveCharacterLimit
		self.repeatedCharacterLimit = {
			if style == .nice && repeatedCharacterLimit != 1 {
				return nil
			}
			return consecutiveCharacterLimit
		}()
		
		// Calculate the number of characters and splits such that the resulting length satisfies the specified minimum length
		let groupSize = {
			switch style {
				case .random: return 3
				case .nice: return 6
			}
		}()
		var numberOfSplits = 0
		var numberOfCharacters = groupSize
		var split = false
		if grouped {
			let splitSize = groupSize + 1
			numberOfSplits = (minLength - numberOfCharacters + splitSize - 1) / splitSize
			numberOfCharacters += numberOfSplits * groupSize
			split = numberOfSplits > 0
		}
		else {
			numberOfCharacters = minLength
		}
		self.numberOfCharacters = numberOfCharacters
		self.group = split ? (groupSize, groupSeparator) : nil
		
		// For each separator inserted due to splitting, remove a character set that requires the separator character
		var requiredCharacterSets = requiredCharacterSets
		requiredCharacterSets.indices
			.filter { requiredCharacterSets[$0].contains(groupSeparator) }
			.prefix(numberOfSplits)
			.reversed()
			.forEach { requiredCharacterSets.remove(at: $0) }
		// Remove any required character not present in the allowed characters
		requiredCharacterSets = requiredCharacterSets
			.map { $0.intersection(allowedCharacters) }
			.filter { !$0.isEmpty }
		self.characterPool = CharacterSet(union: requiredCharacterSets)
		self.requiredCharacterSets = requiredCharacterSets
		
		// If we have more requirements of the type "need a character from set" than the length of the password we want to generate, then
		// we will never be able to meet these requirements, and we'll end up in an infinite loop generating passwords. To avoid this,
		// reset required character sets if the requirements are impossible to meet.
		precondition(requiredCharacterSets.count <= numberOfCharacters, "Unable to meet requirements: More required character sets (\(requiredCharacterSets.count)) specified than number of password characters (\(numberOfCharacters)) to generate")
	}
	
	
	private func randomInt(max: Int) -> Int {
		Int.random(in: 0..<max)
	}
	private func randomInt(range: Range<Int>) -> Int {
		Int.random(in: range)
	}
	private func randomInt(range: ClosedRange<Int>) -> Int {
		Int.random(in: range)
	}
	
	private func randomCharacter<C: CharacterCollection>(in collection: C) -> Character {
		collection.randomElement()!
	}
	
	private func randomConsonant() -> [Character] {
		[randomCharacter(in: CharacterSet.unambiguousLowerConsonants)]
	}
	
	private func randomVowel() -> [Character] {
		[randomCharacter(in: CharacterSet.unambiguousLowerVowels)]
	}
	
	private func randomDigit() -> [Character] {
		[randomCharacter(in: CharacterSet.digit)]
	}
	
	private func randomSyllable() -> [Character] {
		randomConsonant() + randomVowel() + randomConsonant()
	}
	
	private func randomWord() -> [Character] {
		randomSyllable() + randomSyllable()
	}
	
	
	private func moreTypeablePassword(numberOfMinimumCharacters: Int) -> [Character] {
		var components = [[Character]]()
		
		// Generate enough words to satisfy minimum number of characters
		let wordLength = 6
		let numberOfNeededCharacters = max(numberOfMinimumCharacters - wordLength, 0)
		let numberOfAdditionalWords = (numberOfNeededCharacters + wordLength - 1) / wordLength
		for _ in 0..<numberOfAdditionalWords {
			components.append(randomWord())
		}
		
		// Password always includes one word starting or ending with a digit
		var digitWord = randomSyllable() + randomConsonant() + randomVowel()
		
		// Insert digit word while ensuring that password doesn't start with a digit
		let digitPosition = randomInt(range: 1...(numberOfAdditionalWords*2 + 1))
		let digitIndex = (digitPosition & 0b1 == 0) ? digitWord.startIndex : digitWord.endIndex
		let digitWordIndex = digitPosition / 2
		digitWord.insert(contentsOf: randomDigit(), at: digitIndex)
		components.insert(digitWord, at: digitWordIndex)
		
		// Uppercase a random character that is not an "o" or a digit
		var password = Array(components.joined())
		let length = password.count
		while true {
			let index = randomInt(max: length)
			let lowercaseChar = password[index]
			if lowercaseChar == "o" || CharacterSet.digit.contains(lowercaseChar) {
				continue
			}
			
			let uppercaseChars = lowercaseChar.uppercased()
			password.replaceSubrange(index...index, with: uppercaseChars)
			
			return password
		}
	}
	
	private func classicPassword(numberOfRandomCharacters: Int, from characterPool: CharacterSet) -> [Character] {
		var randomCharArray = [Character]()
		for _ in 0..<numberOfRandomCharacters {
			randomCharArray.append(randomCharacter(in: characterPool))
		}
		return randomCharArray
	}
	
	private func splitArray<T>(_ array: [T], separator: T, groupSize: Int) -> [T] {
		var components: [ArraySlice<T>] = []
		var startIndex = array.startIndex
		while startIndex < array.endIndex {
			let endIndex = min(startIndex + groupSize, array.endIndex)
			components.append(array[startIndex..<endIndex])
			startIndex = endIndex
		}
		return Array(components.joined(separator: [separator]))
	}
	
	
	private func passwordHasNotExceededConsecutiveCharLimit(password: [Character], consecutiveCharLimit: Int) -> Bool {
		let passwordUnicodeScalars = password.flatMap(\.unicodeScalars)
		var longestConsecutiveCharLength = 1
		var firstConsecutiveCharIndex = 0
		// Both "123" or "abc" and "321" or "cba" are considered consecutive.
		var isSequenceAscending: Bool?
		for i in 1..<passwordUnicodeScalars.count {
			let currCharCode = passwordUnicodeScalars[i].value
			let prevCharCode = passwordUnicodeScalars[i-1].value
			if let _isSequenceAscending = isSequenceAscending {
				// If `isSequenceAscending` is not nil, then we know that we are in the middle of an existing
				// pattern. Check if the pattern continues based on whether the previous pattern was
				// ascending or descending.
				if (_isSequenceAscending && currCharCode == prevCharCode + 1) || (!_isSequenceAscending && currCharCode == prevCharCode - 1) {
					continue
				}
				
				// Take into account the case when the sequence transitions from descending
				// to ascending.
				if currCharCode == prevCharCode + 1 {
					firstConsecutiveCharIndex = i - 1
					isSequenceAscending = true
					continue
				}
				
				isSequenceAscending = nil
			} else if currCharCode == prevCharCode + 1 {
				isSequenceAscending = true
				continue
			} else if currCharCode == prevCharCode - 1 {
				isSequenceAscending = false
				continue
			}
			
			let currConsecutiveCharLength = i - firstConsecutiveCharIndex
			if currConsecutiveCharLength > longestConsecutiveCharLength {
				longestConsecutiveCharLength = currConsecutiveCharLength
			}
			
			firstConsecutiveCharIndex = i
		}
		
		if isSequenceAscending ?? false {
			let currConsecutiveCharLength = passwordUnicodeScalars.count - firstConsecutiveCharIndex
			if currConsecutiveCharLength > longestConsecutiveCharLength {
				longestConsecutiveCharLength = currConsecutiveCharLength
			}
		}
		
		return longestConsecutiveCharLength <= consecutiveCharLimit
	}
	
	private func passwordHasNotExceededRepeatedCharLimit(password: [Character], repeatedCharLimit: Int) -> Bool {
		var longestRepeatedCharLength = 1
		var lastRepeatedChar = password[0]
		var lastRepeatedCharIndex = 0
		for i in 1..<password.count {
			let currChar = password[i]
			if currChar == lastRepeatedChar {
				continue
			}
			
			let currRepeatedCharLength = i - lastRepeatedCharIndex
			if currRepeatedCharLength > longestRepeatedCharLength {
				longestRepeatedCharLength = currRepeatedCharLength
			}
			
			lastRepeatedChar = currChar
			lastRepeatedCharIndex = i
		}
		return longestRepeatedCharLength <= repeatedCharLimit
	}
	
	private func passwordContainsRequiredCharacters(password: [Character], requiredCharacterSets: [CharacterSet]) -> Bool {
		for requiredCharacterSet in requiredCharacterSets {
			var hasRequiredChar = false
			for char in password {
				if requiredCharacterSet.contains(char) {
					hasRequiredChar = true
					break
				}
			}
			if !hasRequiredChar {
				return false
			}
		}
		return true
	}
	
	
	func generatePassword() -> String {
		while true {
			var password = [Character]()
			
			switch style {
				case .random:
					password = classicPassword(numberOfRandomCharacters: numberOfCharacters, from: characterPool)
					if let group {
						password = splitArray(password, separator: group.separator, groupSize: group.size)
					}
					
					if !passwordContainsRequiredCharacters(password: password, requiredCharacterSets: requiredCharacterSets) {
						continue
					}
					
				case .nice:
					password = moreTypeablePassword(numberOfMinimumCharacters: numberOfCharacters)
					if let group {
						password = splitArray(password, separator: group.separator, groupSize: group.size)
					}
			}
			
			if let repeatedCharacterLimit, repeatedCharacterLimit >= 1, !passwordHasNotExceededRepeatedCharLimit(password: password, repeatedCharLimit: repeatedCharacterLimit) {
				continue
			}
			if let consecutiveCharacterLimit, consecutiveCharacterLimit >= 1, !passwordHasNotExceededConsecutiveCharLimit(password: password, consecutiveCharLimit: consecutiveCharacterLimit) {
				continue
			}
			
			return String(password)
		}
	}
}
